ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/WorkloadTableToolbar.tsx

```diff
@@ -1,11 +1,222 @@
-import React from 'react';
+import React, { useEffect } from 'react';
+import {
+    Toolbar,
+    ToolbarItem,
+    ToolbarFilter,
+    ToolbarToggleGroup,
+    ToolbarGroup,
+    ToolbarContent,
+    ToolbarChip,
+    Flex,
+} from '@patternfly/react-core';
+import { FilterIcon } from '@patternfly/react-icons';
+import { Globe } from 'react-feather';
+
+import useURLSearch from 'hooks/useURLSearch';
+import { uniq } from 'lodash';
+import { DefaultFilters, VulnerabilitySeverityLabel, FixableStatus } from './types';
+import FilterResourceDropdown, { Resource } from './FilterResourceDropdown';
+import FilterAutocompleteSelect from './FilterAutocompleteSelect';
+import CVESeverityDropdown from './CVESeverityDropdown';
+import CVEStatusDropdown from './CVEStatusDropdown';
+
+type FilterType = 'resource' | 'Severity' | 'Fixable';
+
+type WorkloadTableToolbarProps = {
+    defaultFilters: DefaultFilters;
+    resourceContext?: Resource;
+};
+
+function WorkloadTableToolbar({ defaultFilters, resourceContext }: WorkloadTableToolbarProps) {
+    const { searchFilter, setSearchFilter } = useURLSearch();
```

Comment:
ideally i'd like to be able to specify the search filters used on this page more specifically than the `string | string[] | undefined` it has here -- any ideas on how to accomplish that?

Reply:
Hmm this is tricky but I agree it would be very useful. I did something similar with just a single parameter name + array of string values [here](https://github.com/stackrox/stackrox/pull/5111/files#diff-fec9b2b0f852150eb19ed6e8eda2b8909a3ccbc6b841e07f19be4839eb14bd8e), but this is a much more complex case.

It seems like the ideal scenario would be to create a new hook that could work like:
```typescript
function useRestrictedURLSearch<SearchType extends SearchFilter>(): {
    searchFilter: SearchType;
    setSearchFilter: (next: SearchType) => void;
} {
    const { searchFilter, setSearchFilter } = useURLSearch();
    // Magic...?
    return { safeSearchFilter, setSafeSearchFilter };
}

export type VulnerabilitySeverityLabel = 'Critical' | 'Important' | 'Moderate' | 'Low';
export type FixableStatus = 'Fixable' | 'Not fixable';

export type DefaultFilters = {
    Severity: VulnerabilitySeverityLabel[];
    Fixable: FixableStatus[];
};

const { searchFilter, setSearchFilter } = useRestrictedURLSearch<DefaultFilters>();

// `searchFilter` is guaranteed to be a `DefaultFilters`
```
but I'm not sure if that is something we would be able to achieve with TypeScript.

Another option could be to do something more verbose that would allow us to check types at runtime like:
```typescript
const defaultFilterDescriptors = {
    Severity: ['Critical', 'Important', 'Moderate', 'Low'] as const,
    Fixable: ['Fixable', 'Not fixable'] as const,
    CVE: 'string',
    // ...other resources
};

const { searchFilter, setSearchFilter } = useRestrictedURLSearch(defaultFilterDescriptors);

// `searchFilter` is asserted to be `typeof defaultFilterDescriptors`, which should be close to matching the `DefaultFilters` type
```
or even more verbose:
```typescript

const defaultFilterDescriptors = {
    Severity: {
        type: 'stringUnionArray',
        values: ['Critical', 'Important', 'Moderate', 'Low'],
        default: undefined,
        optional: true,
    },
    Fixable: {
        type: 'stringUnionArray',
        values: ['Fixable', 'Not fixable'],
        default: undefined,
        optional: true,
    },
    CVE: {
        type: 'stringArray',
        optional: true,
        // Any string value is valid
    },
    // ...other resources
} as const;

const { searchFilter, setSearchFilter } = useRestrictedURLSearch(defaultFilterDescriptors);
```
but I think the latter two would require some serious type gymnastics to get the result we want in the general case.

We might be able to get this to work with something like [Zod](https://github.com/colinhacks/zod) too, but I haven't used it enough to know for sure.